skfem/utils.py

```diff
@@ -219,11 +221,87 @@ def _flatten_helper(S, key):
     raise NotImplementedError("Unable to flatten the given set of DOFs.")
 
 
+def enforce(A: spmatrix,
+            b: Optional[ndarray] = None,
+            x: Optional[ndarray] = None,
+            I: Optional[DofsCollection] = None,
+            D: Optional[DofsCollection] = None,
+            diag: float = 1.) -> EnforcedSystem:
+    r"""Enforce degrees-of-freedom of a linear system.
+
+    .. note::
+
+        The original system is both modified (for performance) and returned
+        (for compatibility with :func:`skfem.utils.solve`).
+
+    Parameters
+    ----------
+    A
+        The system matrix
+    b
+        Optionally, the right hand side vector.
```

Comment:
For `condense`, the second positional argument, `b` is taken to be the mass matrix of a generalized algebraic eigenvalue problem if it's an `spmatrix` rather than an `ndarray`.  Is it worth doing the same here?  It will be the same entries that are to be affected in both matrices.

Reply:
I think of the generalized algebraic eigenvalue problem as arising from the dynamical system Mu' = Ku by the separation of variables u(t) = exp(st) v as sMv=Kv.  The steady problem is Kv=0 and that's the one that takes the possibly non-zero Dirichlet values so it's K that gets ones on the diagonal. In rhe unsteady problem, the Dirichlet degrees of freedom don't change so the corresponding diagonal entries of M should be zero. In the generalized eigenvalue problem, each Dirichlet degree of freedom has a s=1/0 = &infin; generalized eigenvalue corresponding to a zero time-constant in exp st = exp (t/0). If the Dirichlet data were changing in time, the degrees of freedom would follow instantaneously, with no relaxation time. Obviating these additional eigenvalues is one motivation for condensation over enforcement.

Reply:
> I think for eigenvalue problems you cannot set inhomogeneous boundary values

No, that's right. The inhomogeneous data would be applied to the steady problem then the unsteady problem rewritten for the perturbation. I skiipper that step in the previous. So Mx' = Kx - b with steady problem Kx=b and perturbation u satisfying the homogeneous ordinary differential equation Mu' = Ku.

Reply:
Can't get this to work, maybe I'm missing something.

Reply:
```python
    if b is not None:
        if isinstance(b, spmatrix):
            # eigenvalue problem
            b = enforce(b, D=D, diag=0.)
        else:
            # set rhs to the given value
            b[D] = x[D]
        return A, b
```

Reply:
https://scicomp.stackexchange.com/questions/31140/dirichlet-boundary-conditions-in-generalized-eigenvalue-problem

Reply:
Seems I was accidentally using `eigsh` which is for symmetric cases,

Reply:
Sorry, I didn't get to put together the MWE before Easter. The string on a MeshLine would be fine or actually even simpler the thermal problem, the one-dimensional eigenvalue analogue of [ex19](https://github.com/kinnala/scikit-fem/blob/3369c040005c90e7edd925694620c8bfd1048286/docs/examples/ex19.py). 

In the scicomp SE question, I think the OP's idea that 1 (=condense) makes most sense for generalized algebraic eigenproblems is right while 2 (=enforce) for &lsquo;solving linear systems of equations&rsquo; is reasonable and matches #591 , although I'm wondering whether its real advantages will shine for unsteady and nonlinear problems. I don't like 3 (a kind of inverse enforce) as it's unphysical. The additional zeros eigenvalues obtained with `enforce` do have a physical interpretation; comparison with (the as yet unwritten) `penalize` should reveal this. SB's suggestion of further modifying the mass matrix to artificially preserve its properties of symmetry and positive definiteness is definitely something that i would rather avoid. It's true that there is a lot of that in the spectral methods and hydrodynamic stability literature but compare how simple [ex29](https://github.com/kinnala/scikit-fem/blob/3369c040005c90e7edd925694620c8bfd1048286/docs/examples/ex29.py) is, using `condense`.

The issue of hanging on to these properties might not be relevant if, as usual in Krylov methods for eigenvalue problems, one is employing a shift (`sigma`) since that modifies the pencil, but that would have to be checked.

Meanwhile, so as not to hold up `enforce` for the static case while the generalized eigenproblem is pondered, particularly since `condense` is possibly preferred there anyway, it might be reasonable to leave the `isinstance(b, spmatrix)` case as `NotImplemented` to begin with.

Reply:
I think I was able to get it to work. I changed `TestEx21` to run using `enforce` and the result was almost the same, maybe 0.1% difference due to unsymmetric matrix and using a different function of ARPACK.